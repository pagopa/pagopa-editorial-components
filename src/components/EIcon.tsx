import * as MuiIcons from '@mui/icons-material';
import { type SvgIconProps } from '@mui/material';

export interface EIconProps extends SvgIconProps {
  name?: keyof typeof MuiIcons;
}

export const EIcon = ({ name, ...rest }: EIconProps) => {
  const Icon = name && MuiIcons[name];
  return Icon ? <Icon {...rest} /> : null;
};
