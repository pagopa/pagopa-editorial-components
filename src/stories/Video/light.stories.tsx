import { type ComponentStory, type ComponentMeta } from '@storybook/react';
import Video from '../../components/Video';
import { EIcon } from '../../components/EIcon';
import { Stack, Typography } from '@mui/material';

export default {
  title: 'Video/light',
  component: Video,
} as ComponentMeta<typeof Video>;

const Template: ComponentStory<typeof Video> = (args) => <Video {...args} />;

export const Regular = Template.bind({});
Regular.args = {
  theme: 'light',
  autoplay: true,
  title: 'Nam eget dui. Etiam rhoncus Maecenas tempus',
  subtitle:
    'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa.',
  src: 'https://assets.mixkit.co/videos/preview/mixkit-person-typing-on-a-computer-in-detail-4907-large.mp4',
};

export const Reverse = Template.bind({});
Reverse.args = {
  theme: 'light',
  reverse: true,
  autoplay: true,
  title: 'Nam eget dui. Etiam rhoncus Maecenas tempus',
  subtitle:
    'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa.',
  src: 'https://assets.mixkit.co/videos/preview/mixkit-person-typing-on-a-computer-in-detail-4907-large.mp4',
};

export const Full = Template.bind({});
Full.args = {
  theme: 'light',
  src: 'https://assets.mixkit.co/videos/preview/mixkit-person-typing-on-a-computer-in-detail-4907-large.mp4',
  autoplay: true,
  full: true,
  loop: true,
};

export const Error = Template.bind({});
Error.args = {
  theme: 'light',
  full: true,
  fallback: (
    <Stack textAlign="center">
      <EIcon icon="ErrorOutline" />
      <Typography variant="h6">
        Spiacenti il video non Ã¨ disponibile in questo momento
      </Typography>
    </Stack>
  ),
};
