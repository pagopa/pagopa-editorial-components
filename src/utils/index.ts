import React from 'react';
import { type Generic } from 'types/components';

/**
 * this function utility can be used to verify
 * if component's props is a JSX to render
 */
export const isJSX = <T>(arg: T | Generic): arg is Generic =>
  React.isValidElement(arg);
