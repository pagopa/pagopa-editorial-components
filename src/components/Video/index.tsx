import { useEffect, useRef, useState } from 'react';
import { Grid, Link, Stack, Typography } from '@mui/material';
import EContainer from '../EContainer';
import { type Generic, type CommonProps } from '../../types/components';
import { isJSX, useIsVisible } from '../../utils';

export interface VideoProps extends CommonProps {
  /** Video url */
  src?: string;
  /** Video starts automatically as soon is visible */
  autoplay?: boolean;
  loop?: boolean;
  title?: string;
  subtitle?: string;
  /** When true the full layout is used. Default: false */
  full?: boolean;
  /** when the full layout is not used, this props change the layout alignment.
   * Default: false (left aligned layout)
   */
  reverse?: boolean;
  /** String or a custome element to be shown when something goes wrong */
  fallback?: string | Generic;
  /** Play button text */
  playButtonLabel?: string;
}

const Video = (props: VideoProps) => {
  const {
    title,
    subtitle,
    src,
    autoplay = false,
    loop = false,
    full = false,
    reverse = false,
    theme = 'dark',
    fallback = 'Ops! Qualcosa Ã¨ andato storto... Riprova in un secondo momento.',
    playButtonLabel = 'Guarda il video',
  } = props;
  const videoRef = useRef<HTMLVideoElement>(null);
  const isVisible = useIsVisible(videoRef);
  const [error, setError] = useState(false);

  useEffect(() => {
    if (!isVisible) return;
    const startVideoWhenVisible = async () => {
      if (autoplay && isVisible) await play();
    };
    startVideoWhenVisible().catch(console.error);
  }, [isVisible]);

  const play = async (e?: React.MouseEvent) => {
    try {
      e?.preventDefault();
      await videoRef.current?.play();
    } catch (error) {
      console.warn(error);
    }
  };

  const type = src?.split('.').pop() ?? ('mp4' || '');

  const render = () => {
    if (error) {
      return isJSX(fallback) ? (
        fallback
      ) : (
        <Typography variant="h6" color="background.paper" textAlign="center">
          {fallback}
        </Typography>
      );
    }
    return (
      <video
        onContextMenu={(e) => {
          e.preventDefault();
        }}
        controlsList="nodownload"
        playsInline
        ref={videoRef}
        controls
        muted
        loop={loop}
        style={{
          aspectRatio: 1.777,
          borderRadius: '25px',
          overflow: 'hidden',
          width: '100%',
        }}
      >
        <source
          src={src}
          type={`video/${type}`}
          onError={() => {
            setError(true);
          }}
        />
      </video>
    );
  };

  return (
    <EContainer
      spacing={{ xs: 3, md: 16 }}
      py={7}
      alignItems="center"
      direction={reverse ? 'row-reverse' : 'row'}
      background={theme === 'dark' ? 'primary.dark' : 'white'}
    >
      <Grid item xs={12} md={full ? 12 : 6}>
        {render()}
      </Grid>
      {!full && (
        <Grid item md={6}>
          <VideoText title={title} subtitle={subtitle} theme={theme} />
          <Stack direction="row" alignItems="center">
            <Link
              component="button"
              onClick={play}
              sx={{
                fontWeight: 700,
                fontSize: '16px',
                textDecoration: 'none',
                color:
                  theme === 'dark' ? 'primary.contrastText' : 'primary.dark',
              }}
            >
              {playButtonLabel}
            </Link>
          </Stack>
        </Grid>
      )}
    </EContainer>
  );
};

interface VideoTextProps extends CommonProps {
  title?: string;
  subtitle?: string;
}

const VideoText = ({ title, subtitle, theme = 'dark' }: VideoTextProps) => {
  const color = theme === 'dark' ? 'primary.contrastText' : 'text.primary';
  return (
    <>
      {title && (
        <Typography variant="h5" mb={4} color={color}>
          {title}
        </Typography>
      )}
      {subtitle && (
        <Typography paragraph sx={{ fontSize: '16px' }} mb={3} color={color}>
          {subtitle}
        </Typography>
      )}
    </>
  );
};

export default Video;
