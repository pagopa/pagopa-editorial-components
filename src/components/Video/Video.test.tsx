/* eslint-disable prettier/prettier */
import { Grid, Typography } from '@mui/material';
import Video from './index';
import '@testing-library/jest-dom';
import { EIcon } from '../EIcon';
import 'intersection-observer';
import * as Utils from '../../utils/index';

/**
 * @jest-environment jsdom
 */

import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import React from 'react';

it('renders correctly standard dark', () => {
  const tree = renderer
    .create(
      <Video
        autoplay
        src="https://assets.mixkit.co/videos/preview/mixkit-person-typing-on-a-computer-in-detail-4907-large.mp4"
        subtitle="Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa."
        theme="dark"
        title="Nam eget dui. Etiam rhoncus Maecenas tempus"
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly reverse dark ', () => {
  const tree = renderer
    .create(
      <Video
        autoplay
        reverse
        src="https://assets.mixkit.co/videos/preview/mixkit-person-typing-on-a-computer-in-detail-4907-large.mp4"
        subtitle="Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa."
        theme="dark"
        title="Nam eget dui. Etiam rhoncus Maecenas tempus"
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly full dark ', () => {
  const tree = renderer
    .create(
      <Video
        autoplay
        full
        loop
        src="https://assets.mixkit.co/videos/preview/mixkit-person-typing-on-a-computer-in-detail-4907-large.mp4"
        theme="dark"
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly error dark ', () => {
  const tree = renderer
    .create(
      <Video
        fallback={
          <Grid textAlign="center">
            <EIcon color="background.paper" icon="ErrorOutline" />
            <Typography color="background.paper" variant="h6">
              Spiacenti il video non Ã¨ disponibile in questo momento
            </Typography>
          </Grid>
        }
        full
        theme="dark"
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('autplays the video if autoplay is set to true and isVisible is changed', () => {
  const useIsVisibleMock = jest.spyOn(Utils, 'useIsVisible');
  useIsVisibleMock.mockReturnValue(true);
  jest.spyOn(React, 'useEffect');
  const playMock = jest
    .spyOn(window.HTMLMediaElement.prototype, 'play')
    .mockImplementation(jest.fn());

  render(
    <Video
      autoplay
      src="https://assets.mixkit.co/videos/preview/mixkit-person-typing-on-a-computer-in-detail-4907-large.mp4"
      subtitle="Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa."
      theme="dark"
      title="Nam eget dui. Etiam rhoncus Maecenas tempus"
    />
  );

  expect(React.useEffect).toHaveBeenCalledTimes(1);
  expect(playMock).toHaveBeenCalledTimes(1);
});

it('does not autplay the video if autoplay is set to true and isVisible is changed to false', () => {
  const useIsVisibleMock = jest.spyOn(Utils, 'useIsVisible');
  useIsVisibleMock.mockReturnValue(false);
  jest.spyOn(React, 'useEffect');
  const playMock = jest
    .spyOn(window.HTMLMediaElement.prototype, 'play')
    .mockImplementation(jest.fn());

  render(
    <Video
      autoplay
      src="https://assets.mixkit.co/videos/preview/mixkit-person-typing-on-a-computer-in-detail-4907-large.mp4"
      subtitle="Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa."
      theme="dark"
      title="Nam eget dui. Etiam rhoncus Maecenas tempus"
    />
  );
  expect(React.useEffect).toHaveBeenCalledTimes(1);
  expect(playMock).toHaveBeenCalledTimes(0);
});
