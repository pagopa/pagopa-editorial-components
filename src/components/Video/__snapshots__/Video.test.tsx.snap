// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly error dark  1`] = `
<section
  className="MuiBox-root css-17cshki"
>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg MuiContainer-disableGutters css-1p4tas-MuiContainer-root"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-spacing-md-16 css-g3cke0-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 css-1idn90j-MuiGrid-root"
      >
        <video
          controls={true}
          controlsList="nodownload"
          loop={false}
          muted={true}
          onContextMenu={[Function]}
          playsInline={true}
          style={
            {
              "aspectRatio": 1.777,
              "borderRadius": "25px",
              "overflow": "hidden",
              "width": "100%",
            }
          }
        >
          <source
            onError={[Function]}
            type="video/mp4"
          />
        </video>
      </div>
    </div>
  </div>
</section>
`;

exports[`renders correctly full dark  1`] = `
<section
  className="MuiBox-root css-17cshki"
>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg MuiContainer-disableGutters css-1p4tas-MuiContainer-root"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-spacing-md-16 css-g3cke0-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 css-1idn90j-MuiGrid-root"
      >
        <video
          controls={true}
          controlsList="nodownload"
          loop={true}
          muted={true}
          onContextMenu={[Function]}
          playsInline={true}
          style={
            {
              "aspectRatio": 1.777,
              "borderRadius": "25px",
              "overflow": "hidden",
              "width": "100%",
            }
          }
        >
          <source
            onError={[Function]}
            src="https://assets.mixkit.co/videos/preview/mixkit-person-typing-on-a-computer-in-detail-4907-large.mp4"
            type="video/mp4"
          />
        </video>
      </div>
    </div>
  </div>
</section>
`;

exports[`renders correctly reverse dark  1`] = `
<section
  className="MuiBox-root css-17cshki"
>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg MuiContainer-disableGutters css-1p4tas-MuiContainer-root"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-spacing-md-16 MuiGrid-direction-xs-row-reverse css-1eju3oe-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1fyyp8j-MuiGrid-root"
      >
        <video
          controls={true}
          controlsList="nodownload"
          loop={false}
          muted={true}
          onContextMenu={[Function]}
          playsInline={true}
          style={
            {
              "aspectRatio": 1.777,
              "borderRadius": "25px",
              "overflow": "hidden",
              "width": "100%",
            }
          }
        >
          <source
            onError={[Function]}
            src="https://assets.mixkit.co/videos/preview/mixkit-person-typing-on-a-computer-in-detail-4907-large.mp4"
            type="video/mp4"
          />
        </video>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
      >
        <h5
          className="MuiTypography-root MuiTypography-h5 css-13kf1b5-MuiTypography-root"
        >
          Nam eget dui. Etiam rhoncus Maecenas tempus
        </h5>
        <p
          className="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph css-1rctpy4-MuiTypography-root"
        >
          Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa.
        </p>
        <div
          className="MuiStack-root css-1d9cypr-MuiStack-root"
        >
          <button
            className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways MuiLink-button css-1d8rg46-MuiTypography-root-MuiLink-root"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
          >
            Guarda il video
          </button>
        </div>
      </div>
    </div>
  </div>
</section>
`;

exports[`renders correctly standard dark 1`] = `
<section
  className="MuiBox-root css-17cshki"
>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg MuiContainer-disableGutters css-1p4tas-MuiContainer-root"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-spacing-md-16 css-g3cke0-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1fyyp8j-MuiGrid-root"
      >
        <video
          controls={true}
          controlsList="nodownload"
          loop={false}
          muted={true}
          onContextMenu={[Function]}
          playsInline={true}
          style={
            {
              "aspectRatio": 1.777,
              "borderRadius": "25px",
              "overflow": "hidden",
              "width": "100%",
            }
          }
        >
          <source
            onError={[Function]}
            src="https://assets.mixkit.co/videos/preview/mixkit-person-typing-on-a-computer-in-detail-4907-large.mp4"
            type="video/mp4"
          />
        </video>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
      >
        <h5
          className="MuiTypography-root MuiTypography-h5 css-13kf1b5-MuiTypography-root"
        >
          Nam eget dui. Etiam rhoncus Maecenas tempus
        </h5>
        <p
          className="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph css-1rctpy4-MuiTypography-root"
        >
          Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa.
        </p>
        <div
          className="MuiStack-root css-1d9cypr-MuiStack-root"
        >
          <button
            className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways MuiLink-button css-1d8rg46-MuiTypography-root-MuiLink-root"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
          >
            Guarda il video
          </button>
        </div>
      </div>
    </div>
  </div>
</section>
`;
